proxy/handler_test.go

```diff
@@ -0,0 +1,60 @@
+package proxy_test
+
+import (
+	"context"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/ONSdigital/dp-legacy-cache-proxy/config"
+	"github.com/ONSdigital/dp-legacy-cache-proxy/proxy"
+	"github.com/gorilla/mux"
+	. "github.com/smartystreets/goconvey/convey"
+)
+
+var babbageEndpoint = "/economy/environmentalaccounts/bulletins/ukenvironmentalaccounts/2013-06-26"
+
+func TestProxyHandler(t *testing.T) {
+	Convey("Given a Proxy Handler ", t, func() {
+		router := mux.NewRouter()
+		ctx := context.Background()
+
+		Convey("when sending a proxy request to Babbage", func() {
+			babbageMockServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusOK)
+				w.Write([]byte("Mock response from BabbageURL"))
+			}))
+			r := httptest.NewRequest(http.MethodGet, babbageEndpoint, http.NoBody)
+			w := httptest.NewRecorder()
+
+			cfg := &config.Config{BabbageURL: babbageMockServer.URL}
+			legacyProxy := proxy.Setup(ctx, router, cfg)
+
+			legacyProxy.Router.ServeHTTP(w, r)
+
+			Convey("Then return the Babbage response with status code 200", func() {
+				So(w.Code, ShouldEqual, http.StatusOK)
+				So(w.Body.String(), ShouldEqual, "Mock response from BabbageURL")
+			})
+		})
+
+		Convey("when there is an error with sending the proxy request to Babbage", func() {
```

Comment:
```suggestion
		Convey("When there is an error with sending the proxy request to Babbage", func() {
```

Reply:
Also, this message is misleading. What the message says would actually be a good test, but that's not what you're actually testing. You're testing that when Babbage returns a 500 error, the proxy also returns that. But this is simply an extension of the test that you've already written above. What you should be testing here is the different scenarios in which the proxy itself throws an error (for instance, when `http.NewRequestWithContext` or `client.Do(proxyReq)` or `io.Copy` fail).

Reply:
Testing `http.NewRequestWithContext` and `io.Copy` turned out to be difficult as we couldn't get them to error through the tests. The only other option we thought of is to mock the http client for `http.NewRequestWithContext ` and for `io.Copy` to simulate a custom Writer that returns an error and would cause it to fail. However both of these seemed a bit unclear and convoluted, would you have any suggestions?

Reply:
I think if some of these errors are hard to reproduce, we're OK not covering them, as they're not really part of the business logic anyway. They're there just as a safeguard.